packages/scratch-vm/src/extensions/scratch3_spotify/index.ts

```diff
@@ -0,0 +1,696 @@
+        }
+        for (let i = 0; i < this.barTimeouts.length; i++) {
+            clearTimeout(this.barTimeouts[i]);
+        }
+    }
+
+    resetTrackData() {
+        this.player = new Tone.Player().toMaster();
+        this.currentArtistName = "no artist";
+        this.currentTrackName = "no track";
+        this.currentAlbumName = "no album";
+        this.trackTempo = 0;
+    }
+
+    requestSearch(query: string) {
+        return new Promise<void> ((resolve, reject) => {
+            if (this.player) {
+                this.player.stop();
+                this.clearTimeouts();
+            }
+
+            if (query == "") {
+                return reject();
+            }
+
+            this.currentBeatNum = 0;
+
+            // if we are making the exact same query again, abort
+            // keeping the same metadata, no need to reload
+            if (query == this.prevQuery) {
+                return resolve();
+            }
+
+            nets(
+                {
+                    url:
+                        "https://api.spotify.com/v1/search?q=" +
+                        query +
+                        "&type=track",
+                    headers: {
+                        Authorization: "Bearer " + this.spotifyToken.value,
+                    },
+                    timeout: SERVER_TIMEOUT,
+                },
+                (err, res, body) => {
+                    if (err) {
+                        console.log("Error with Spotify API query");
+                        console.log(err);
+                        console.log(res);
+                        return reject();
+                    }
+
+                    if (res.statusCode !== 200) {
+                        console.log("Error with Spotify API query");
+                        console.log(res);
+                        console.log(JSON.parse(body));
+
+                        // if the error is a 401 (unauthorized), the token probably has expired, so request a new one.
+                        // TODO make sure I'm doing this right
+                        if (res.statusCode == 401) {
+                            // TODO figure out why this isn't calling refresh access token
+                            console.log("401 error");
+                            getAccessToken().then((token) => {
+                                this.spotifyToken = token;
+                                return reject();
+                            });
+                        } else {
+                            console.error(
+                                "Spotify token error: " + res.statusCode
+                            );
+                            return reject();
+                        }
+                    }
+
+                    // success
+                    this.prevQuery = query;
+                    let trackObjects = JSON.parse(body).tracks.items;
+
+                    // fail if there are no tracks
```

Comment:
This might be coming from the old implementation, but I think it's great practice to capture information like this in a variable, since it's easy for comments to get out of date / aren't always as easily parsable at a glance. 

So instead, I'd recommend:
```ts 
noTracksFound = !trackObjects || trackObjects.length === 0;
if (noTracksFound) {
   ...
}
```